### PR TITLE
atomic_compare_exchange: compare fcn name typo

### PR DESCRIPTION
**Pull request by ndellingwood**

### CODE REVIEW

**Comment by ndellingwood** [comment]
This showed up in nightly job with uvm and rdc and deprecated code enabled

**Review by dalg24 — DISMISSED** [review]

**Comment by dalg24** [comment]
We might want to add coverage for this :/

**Review by masterleinad — APPROVED** [review]

**Comment by ndellingwood** [comment]
@dalg24 this was picked up with a kokkos-kernels job, cuda/11.2.2+uvm+rdc+deprecated code enabled on Volta70

A kokkos job with cuda/11.8.0+uvm+deprecated code enabled on Hopper did not pick up the error

I'm not sure which difference is the trigger - rdc, arch, or cuda version, is it clear from the desul wrapper code what config option is needed to trigger?

**Comment by lucbv** [comment]
I guess we are not really interested in setting up a test that checks that this deprecated feature is working properly. On the other hand it would have avoided the current issue...

**Review by romintomasetti — APPROVED** [review]

**Comment by masterleinad** [comment]
> I guess we are not really interested in setting up a test that checks that this deprecated feature is working properly. On the other hand it would have avoided the current issue...

You are free to propose appropriate changes to our CI.

File: core/unit_test/TestAtomics.hpp [hunk 1]
```diff
@@ -308,32 +265,70 @@ T CASLoopSerial(int loop) {
...
 
   KOKKOS_INLINE_FUNCTION
-  void operator()(int i) const {
-    T old = Kokkos::atomic_exchange(&data(), (T)i);
-    Kokkos::atomic_fetch_add(&data2(), old);
+  void operator()(int) const {
+    T newval, assumed;
+
+    do {
+      assumed = Kokkos::volatile_load(&data());
```
Comment (dalg24) [inline core/unit_test/TestAtomics.hpp:287]: @crtrott do you have a better idea?

File: core/unit_test/TestAtomicOperations.hpp [hunk 2]
```diff
@@ -277,6 +277,63 @@ struct LoadStoreAtomicTest {
   static const char* name() { return "load/store"; }
 };
 
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+#ifdef KOKKOS_ENABLE_DEPRECATION_WARNINGS
+KOKKOS_IMPL_DISABLE_DEPRECATED_WARNINGS_PUSH()
+#endif
+struct DeprecatedAssignAtomicTest {
```
Comment (dalg24) [inline core/unit_test/TestAtomicOperations.hpp:284]: Is a copy/paste from `LoadStoreAtomicTest` with atomic_load -> atomic_assign

File: core/unit_test/TestAtomicOperations.hpp [hunk 3]
```diff
@@ -277,6 +277,63 @@ struct LoadStoreAtomicTest {
...
+    return Kokkos::pair<T, T>(old_val, update);
+  }
+  template <class T>
+  KOKKOS_FUNCTION static T op(T, T update) {
+    return update;
+  }
+  static const char* name() { return "load/assign"; }
+};
+
+struct DeprecatedIncrementAtomicTest {
```
Comment (dalg24) [inline core/unit_test/TestAtomicOperations.hpp:301]: Are copy paste from `{Inc,Dec}AtomicTest` but with the atomic "reduce" operation that uses there deprecated version.  They are no fetch counterparts to test

File: core/unit_test/TestAtomicOperations.hpp [hunk 4]
```diff
@@ -472,11 +529,28 @@ bool AtomicOperationsTestIntegralType(int old_val_in, int update_in, int test) {
                             : true;
 #endif
     case 13:
-      return atomic_op_test<IncAtomicTest, T, ExecSpace>(old_val, update);
+      return
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+          atomic_op_test<DeprecatedIncrementAtomicTest, T, ExecSpace>(old_val,
+                                                                      update) &&
+#endif
+          atomic_op_test<IncAtomicTest, T, ExecSpace>(old_val, update);
```
Comment (dalg24) [inline core/unit_test/TestAtomicOperations.hpp:537]: Didn't want to mess with the driver so require both to be true

File: core/unit_test/TestAtomics.hpp [hunk 5]
```diff
@@ -157,17 +157,6 @@ struct AddFunctor {
   void operator()(int) const { Kokkos::atomic_fetch_add(&data(), (T)1); }
 };
 
-template <class T, class DEVICE_TYPE>
-struct AddFunctorReduce {
```
Comment (dalg24) [inline core/unit_test/TestAtomics.hpp:161]: Reduce are pointless and do the exact same thing as the parallel fors

**Comment by ndellingwood** [comment]
Testing-wise the initial error (in function name) is also showing up in Trilinos builds, which disable deprecated code by default, so I added the Block Promotion label

File: core/unit_test/TestAtomics.hpp (hunk 1 above)
Comment (crtrott) [reply core/unit_test/TestAtomics.hpp:287]: Agreed `atomic_load` is probably better, however we could also see that as an integral part of testing volatile_load and deprecating it?

Comment (crtrott) [reply core/unit_test/TestAtomics.hpp:287]: Yeah I guess I am ok with the volatile-load

**Review by crtrott — APPROVED** [review]

**Comment by ndellingwood** [comment]
The `OPENACC-NVHPC-CUDA-12.2` job fails to compile some of the new tests, are push/pop macros needed to suppress warning?

```
"/var/jenkins/workspace/Kokkos_PR-7467/core/unit_test/TestAtomicOperations.hpp", line 305: error: function "Kokkos::atomic_increment(T *) [with T=int]" was declared deprecated ("Use atomic_inc() instead!") [deprecated_entity_with_custom_message]

      Kokkos::atomic_increment(ptr_op);

              ^

"/var/jenkins/workspace/Kokkos_PR-7467/core/src/Kokkos_Atomics_Desul_Wrapper.hpp", line 134: note: because of a "deprecated" attribute

  template<class T> KOKKOS_DEPRECATED_WITH_COMMENT("Use atomic_inc() instead!") KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_increment(T* ptr) { atomic_inc(ptr); }

                    ^

          detected during:

            instantiation of "auto TestAtomicOperations::DeprecatedIncrementAtomicTest::atomic_op(T *, T *, T *, T) [with T=int]" at line 396

            instantiation of "bool TestAtomicOperations::atomic_op_test<Op,T,ExecSpace>(T, T) [with Op=TestAtomicOperations::DeprecatedIncrementAtomicTest, T=int, ExecSpace=Kokkos::Serial]" at line 535

            instantiation of "bool TestAtomicOperations::AtomicOperationsTestIntegralType<T,ExecSpace>(int, int, int) [with T=int, ExecSpace=Kokkos::Serial]" at line 29 of "/var/jenkins/workspace/Kokkos_PR-7467/core/unit_test/TestAtomicOperations_int.hpp"

Remark: individual warnings can be suppressed with "--diag_suppress <warning-name>"

"/var/jenkins/workspace/Kokkos_PR-7467/core/unit_test/TestAtomicOperations.hpp", line 321: error: function "Kokkos::atomic_decrement(T *) [with T=int]" was declared deprecated ("Use atomic_dec() instead!") [deprecated_entity_with_custom_message]

      Kokkos::atomic_decrement(ptr_op);

              ^

"/var/jenkins/workspace/Kokkos_PR-7467/core/src/Kokkos_Atomics_Desul_Wrapper.hpp", line 135: note: because of a "deprecated" attribute
...
```

**Review by ndellingwood — COMMENTED** [review]
I'm not sure if wrapping the deprecated fcn calls explicitly with warning push/pops helps with the -Werror ?

File: core/unit_test/TestAtomicOperations.hpp [hunk 6]
```diff
@@ -277,6 +277,63 @@ struct LoadStoreAtomicTest {
...
+    return old + 1;
+  }
+  static const char* name() { return "increment"; }
+};
+
+struct DeprecatedDecrementAtomicTest {
+  template <class T>
+  KOKKOS_FUNCTION static auto atomic_op(T* ptr_op, T* ptr_fetch_op,
+                                        T* ptr_op_fetch, T) {
+    Kokkos::atomic_decrement(ptr_op);
```
[the remark this answers is not in the document]
Comment (ndellingwood) [reply core/unit_test/TestAtomicOperations.hpp:321]: Adding a warning PUSH-POP wrapper macro might be helpful for single line usages like this?

**Comment by masterleinad** [comment]
@ndellingwood You'll need
```diff
diff --git a/core/src/Kokkos_Macros.hpp b/core/src/Kokkos_Macros.hpp
index efbedbb89..4e51a6a43 100644
--- a/core/src/Kokkos_Macros.hpp
+++ b/core/src/Kokkos_Macros.hpp
@@ -603,8 +603,9 @@ static constexpr bool kokkos_omp_on_host() { return false; }
 
 // clang-format off
 #if defined(__NVCOMPILER)
-  #define KOKKOS_IMPL_DISABLE_DEPRECATED_WARNINGS_PUSH() \
-    _Pragma("diag_suppress 1216")
+#define KOKKOS_IMPL_DISABLE_DEPRECATED_WARNINGS_PUSH() \
+    _Pragma("diag_suppress 1216") \
+    _Pragma("diag_suppress deprecated_entity_with_custom_message") 
   #define KOKKOS_IMPL_DISABLE_DEPRECATED_WARNINGS_POP() \
     _Pragma("diag_default 1216")
 #elif defined(__EDG__)
```

**Comment by ndellingwood** [comment]
@masterleinad thanks, I'll also revert [2fea7c6](https://github.com/kokkos/kokkos/pull/7467/commits/2fea7c6bae320690ab20feffc485884a51d66193)

**Comment by ndellingwood** [comment]
@masterleinad I added the pragma to push/pop in [8941659](https://github.com/kokkos/kokkos/pull/7467/commits/8941659645a7edd4d68e33a03e74fd317f70ed34)

**Comment by ndellingwood** [comment]
CI `OPENMPTARGET-Clang' tests failed with `1: Error: no GPU available for execution.`, system issue unrelated to this PR

**Review by ldh4 — APPROVED** [review]

**Pull request merged**